modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheAdapter.java

```diff
@@ -3653,6 +3654,52 @@ public void awaitLastFut() {
         }
     }
 
+    /**
+     * Replaces previous async operation future on transaction suspend.
+     */
+    public @Nullable FutureHolder suspendLastFut() {
+        FutureHolder holder = lastFut.get();
+
+        IgniteInternalFuture fut = holder.future();
+
+        if (fut != null && !fut.isDone()) {
+            lastFut.set(new FutureHolder());
+
+            return holder;
+        }
+        else
+            return null;
+    }
+
+    /**
+     * Replaces previous async operation future on transaction resume.
+     */
+    public void resumeLastFut(FutureHolder holder) {
+        IgniteInternalFuture resumedFut = holder.future();
+
+        if (resumedFut == null || resumedFut.isDone())
+            return;
+
+        FutureHolder threadHolder = lastFut.get();
+
+        IgniteInternalFuture threadFut = threadHolder.future();
+
+        if (threadFut != null && !threadFut.isDone()) {
+            threadHolder.lock();
```

Comment:
It seems that we should acquire lock before checking that `threadFut` is not null and is not completed, because `FutureHolder` can be cleared concurrently.

Reply:
Checking just local variable is enough. Future can be completed after lock as well, so additional check is redundant. 
Check for completion also performed by compound future, so here code is safe.